state overrides - nonce should allow hex

It looks to me from those logs like the user has supplied a nonce as a hex value, and besu is trying to deserialize it to a long probably expecting it to be a decimal/int

```
2024-12-08 23:40:06.987+00:00 | vert.x-worker-thread-6 | ERROR | BaseJsonRpcProcessor | Error processing method: eth_call [{"from":"0x148c930352929bc4676214a6333ff3d60c14f3da","to":"0xb4ffef15daf4c02787bc533
2580b838ce39805f5","gas":"0x317ff4","data":"0x70a08231000000000000000000000000148c930352929bc4676214a6333ff3d60c14f3da"},"0xc77281",{"0x148c930352929bc4676214a6333ff3d60c14f3da":{"nonce":"0x9e","balance":"0x
215d7f2a0e82433","movePrecompileToAddress":null},"0x28f6899ff643261ca9766ddc251b359a2d00b945":{"stateDiff":{"0x4c3248b57289b773fb4b4d65000ea6e7586ec5f43b9368c3f5fbe242405ae278":"0x6774858067562e930000008ac72
30489e800000000000374922507579f36ea21","0x74f7ea035dfa10f304cf0b8e7e5d9c0f09e863cd8d54cd9bb93157d81cbf10aa":"0x000000000000000000000000000000000000000000000374922507579f36ea21","0xa8414727a6198ba40a2f5e89741
3cff87745aada8834a9dec6b30698cdd798eb":"0xa1ad504367562e9300000000000000000000000000004a6e6ce87410aa4b5006","0xbb1402ab158438f86c9f2b558da42f68d1d0ec92437cd2f7f075dfb8afecb02c":"0x000000000000000000001d528b6
4b92455e7410000000374922507579f36ea21"},"movePrecompileToAddress":null},"0x2f9bb73a8e98793e26cb2f6c4ad037bdf1c6b269":{"stateDiff":{"0x1fd84a755d3a9c78ff8b3bd554246082782d06efd21cdf78c58b0ad75e892b9f":"0x0000
0000000984c5005034e416ff086d0000000003563d3452ef16b32ef604b8","0x1fd84a755d3a9c78ff8b3bd554246082782d06efd21cdf78c58b0ad75e892ba0":"0x0000000000189c384b74901c987f0e3900000000036721ecedf5f2e10ab2412f","0x1fd8
4a755d3a9c78ff8b3bd554246082782d06efd21cdf78c58b0ad75e892ba1":"0x00000000000000000000020067562e9300000000002c6f7077d2986d050fe1c7","0x1fd84a755d3a9c78ff8b3bd554246082782d06efd21cdf78c58b0ad75e892ba6":"0x0000
0000000000000000000000000000000000000000000000120a58a469bd3a"},"movePrecompileToAddress":null},"0xb4ffef15daf4c02787bc5332580b838ce39805f5":{"stateDiff":{"0x00000000000000000000000000000000000000000000000000
00000000000036":"0x00000000000000000000000000000000000000000000023925601eac764778be","0x6ae29f6c72aa6fa7308c28d7d43c83524fa8c6631a476f705e12ff16dab745cb":"0x0000000003563d3452ef16b32ef604b8000000000000000001
c02dd25ac765a4","0xcaff7ca3c07ca5c9cde048b62c7cfae172a46c9085257f8e8a490f3662a5eb90":"0x0000000003563d3452ef16b32ef604b800000000000000000000000000000000","0xcbad9e80aa2f40218b77a8608899af04d15bd48f0c229d9c22
c9f0955dadc9da":"0x00000000000000000000000000000000000000000000000001ced6f2ea95bc58"},"movePrecompileToAddress":null},"0xe5d7c2a44ffddf6b295a15c148167daaaf5cf34f":{"balance":"0x43537bcbb01f0e039f2","stateDif
f":{"0xb0c6eca2f0dc82fff04395ff787d1ab4c55a8f88b9a85ba295da8a5e0e89f940":"0x0000000000000000000000000000000000000000000000ee0581bf089e16c06f"},"movePrecompileToAddress":null}},{"time":"0x67562e93"}]         
org.hyperledger.besu.ethereum.api.jsonrpc.internal.exception.InvalidJsonRpcRequestException: Invalid account overrides parameter (index 2)  
[9:53](https://consensys.slack.com/archives/C03957X9XKP/p1733702034988039?thread_ts=1733701966.061549&cid=C03957X9XKP)
Caused by: org.hyperledger.besu.ethereum.api.jsonrpc.internal.parameters.JsonRpcParameter$JsonRpcParameterException: Invalid json rpc parameter at index 2. Supplied value was: '{0x148c930352929bc4676214a6333
ff3d60c14f3da={nonce=0x9e, balance=0x215d7f2a0e82433, movePrecompileToAddress=null}, 0x28f6899ff643261ca9766ddc251b359a2d00b945={stateDiff={0x4c3248b57289b773fb4b4d65000ea6e7586ec5f43b9368c3f5fbe242405ae278=
0x6774858067562e930000008ac7230489e800000000000374922507579f36ea21, 0x74f7ea035dfa10f304cf0b8e7e5d9c0f09e863cd8d54cd9bb93157d81cbf10aa=0x000000000000000000000000000000000000000000000374922507579f36ea21, 0xa8
414727a6198ba40a2f5e897413cff87745aada8834a9dec6b30698cdd798eb=0xa1ad504367562e9300000000000000000000000000004a6e6ce87410aa4b5006, 0xbb1402ab158438f86c9f2b558da42f68d1d0ec92437cd2f7f075dfb8afecb02c=0x0000000
00000000000001d528b64b92455e7410000000374922507579f36ea21}, movePrecompileToAddress=null}, 0x2f9bb73a8e98793e26cb2f6c4ad037bdf1c6b269={stateDiff={0x1fd84a755d3a9c78ff8b3bd554246082782d06efd21cdf78c58b0ad75e8
92b9f=0x00000000000984c5005034e416ff086d0000000003563d3452ef16b32ef604b8, 0x1fd84a755d3a9c78ff8b3bd554246082782d06efd21cdf78c58b0ad75e892ba0=0x0000000000189c384b74901c987f0e3900000000036721ecedf5f2e10ab2412f
, 0x1fd84a755d3a9c78ff8b3bd554246082782d06efd21cdf78c58b0ad75e892ba1=0x00000000000000000000020067562e9300000000002c6f7077d2986d050fe1c7, 0x1fd84a755d3a9c78ff8b3bd554246082782d06efd21cdf78c58b0ad75e892ba6=0x0
0000000000000000000000000000000000000000000000000120a58a469bd3a}, movePrecompileToAddress=null}, 0xb4ffef15daf4c02787bc5332580b838ce39805f5={stateDiff={0x00000000000000000000000000000000000000000000000000000
00000000036=0x00000000000000000000000000000000000000000000023925601eac764778be, 0x6ae29f6c72aa6fa7308c28d7d43c83524fa8c6631a476f705e12ff16dab745cb=0x0000000003563d3452ef16b32ef604b8000000000000000001c02dd25a
c765a4, 0xcaff7ca3c07ca5c9cde048b62c7cfae172a46c9085257f8e8a490f3662a5eb90=0x0000000003563d3452ef16b32ef604b800000000000000000000000000000000, 0xcbad9e80aa2f40218b77a8608899af04d15bd48f0c229d9c22c9f0955dadc9
da=0x00000000000000000000000000000000000000000000000001ced6f2ea95bc58}, movePrecompileToAddress=null}, 0xe5d7c2a44ffddf6b295a15c148167daaaf5cf34f={balance=0x43537bcbb01f0e039f2, stateDiff={0xb0c6eca2f0dc82ff
f04395ff787d1ab4c55a8f88b9a85ba295da8a5e0e89f940=0x0000000000000000000000000000000000000000000000ee0581bf089e16c06f}, movePrecompileToAddress=null}}' of type: 'java.util.LinkedHashMap' - expected type: 'org.
hyperledger.besu.ethereum.util.AccountOverrideMap'                                                                                                                                                             
    at org.hyperledger.besu.ethereum.api.jsonrpc.internal.parameters.JsonRpcParameter.optional(JsonRpcParameter.java:78) 
[9:54](https://consensys.slack.com/archives/C03957X9XKP/p1733702044994839?thread_ts=1733701966.061549&cid=C03957X9XKP)
Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Cannot deserialize value of type `java.lang.Long` from String "0x9e": not a valid `java.lang.Long` value                                 
 at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 56] (through reference chain: org.hyperledger.besu.ethereum.util.AccountOverrideMap["0x148c930352929bc4676214
a6333ff3d60c14f3da"]->org.hyperledger.besu.ethereum.util.AccountOverride$Builder["nonce"])                                                                                                                     
    at com.fasterxml.jackson.databind.exc.InvalidFormatException.from(InvalidFormatException.java:67)  
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,7 +20,6 @@
 
 ### Bug fixes
 - Correct default parameters for frontier transactions in `eth_call` and `eth_estimateGas` [#7965](https://github.com/hyperledger/besu/pull/7965)
-- Correctly parse nonce as hex in `eth_call` account overrides [#7999](https://github.com/hyperledger/besu/pull/7999)
 
 ## 24.12.0
 
@@ -74,6 +73,7 @@
 - Fix registering new metric categories from plugins [#7825](https://github.com/hyperledger/besu/pull/7825)
 - Fix CVE-2024-47535 [7878](https://github.com/hyperledger/besu/pull/7878)
 - Fix QBFT prepared block based proposal validation [#7875](https://github.com/hyperledger/besu/pull/7875)
+- Correctly parse nonce as hex in `eth_call` account overrides [#7999](https://github.com/hyperledger/besu/pull/7999)
 
 ## 24.10.0
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,7 +19,8 @@
 - Proper support for `pending` block tag when calling `eth_estimateGas` and `eth_createAccessList` [#7951](https://github.com/hyperledger/besu/pull/7951)
 
 ### Bug fixes
--  Correct default parameters for frontier transactions in `eth_call` and `eth_estimateGas` [#7965](https://github.com/hyperledger/besu/pull/7965)
+- Correct default parameters for frontier transactions in `eth_call` and `eth_estimateGas` [#7965](https://github.com/hyperledger/besu/pull/7965)
+- Correctly parse nonce as hex in `eth_call` account overrides [#7999](https://github.com/hyperledger/besu/pull/7999)
 
 ## 24.12.0
 

--- a/datatypes/src/main/java/org/hyperledger/besu/datatypes/AccountOverride.java
+++ b/datatypes/src/main/java/org/hyperledger/besu/datatypes/AccountOverride.java
@@ -21,6 +21,7 @@ import java.util.Optional;
 import com.fasterxml.jackson.annotation.JsonAnySetter;
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
 import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import org.apache.tuweni.bytes.Bytes;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -106,11 +107,11 @@ public class AccountOverride {
     /**
      * Sets the nonce override
      *
-     * @param nonce the nonce override
+     * @param nonce the nonce override in hex
      * @return the builder
      */
-    public Builder withNonce(final Long nonce) {
-      this.nonce = Optional.ofNullable(nonce);
+    public Builder withNonce(final String nonce) {
+      this.nonce = Optional.of(Bytes.fromHexStringLenient(nonce).toLong());
       return this;
     }
 

--- a/ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthCallTest.java
+++ b/ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthCallTest.java
@@ -104,7 +104,7 @@ public class EthCallTest {
   @Test
   public void someAccountOverrides() {
     AccountOverrideMap expectedOverrides = new AccountOverrideMap();
-    AccountOverride override = new AccountOverride.Builder().withNonce(88L).build();
+    AccountOverride override = new AccountOverride.Builder().withNonce("0x9e").build();
     final Address address = Address.fromHexString("0xd9c9cd5f6779558b6e0ed4e6acf6b1947e7fa1f3");
     expectedOverrides.put(address, override);
 

--- a/ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGasTest.java
+++ b/ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthEstimateGasTest.java
@@ -116,7 +116,7 @@ public class EthEstimateGasTest {
   @Test
   public void someAccountOverrides() {
     AccountOverrideMap expectedOverrides = new AccountOverrideMap();
-    AccountOverride override = new AccountOverride.Builder().withNonce(88L).build();
+    AccountOverride override = new AccountOverride.Builder().withNonce("0x9e").build();
     final Address address = Address.fromHexString("0xd9c9cd5f6779558b6e0ed4e6acf6b1947e7fa1f3");
     expectedOverrides.put(address, override);
 

--- a/ethereum/core/src/test/java/org/hyperledger/besu/ethereum/util/AccountOverrideParameterTest.java
+++ b/ethereum/core/src/test/java/org/hyperledger/besu/ethereum/util/AccountOverrideParameterTest.java
@@ -52,7 +52,7 @@ public class AccountOverrideParameterTest {
             + "\":"
             + "{"
             + "\"balance\": \"0x01\","
-            + "\"nonce\": 88"
+            + "\"nonce\": \"0x9e\""
             + "}}],\"id\":1}";
 
     final JsonRpcRequestContext request = new JsonRpcRequestContext(readJsonAsJsonRpcRequest(json));
@@ -62,7 +62,7 @@ public class AccountOverrideParameterTest {
     final AccountOverride accountOverride =
         accountOverrideParam.get(Address.fromHexString(ADDRESS_HEX1));
 
-    assertThat(accountOverride.getNonce()).isEqualTo(Optional.of(88L));
+    assertThat(accountOverride.getNonce().get()).isEqualTo(158);
     assertThat(accountOverride.getBalance()).isEqualTo(Optional.of(Wei.of(1)));
     assertFalse(accountOverride.getStateDiff().isPresent());
   }
@@ -97,6 +97,34 @@ public class AccountOverrideParameterTest {
   }
 
   @Test
+  public void jsonWithHexNonceDeserializesCorrectly() throws Exception {
+    final String json =
+        "{\"jsonrpc\":\"2.0\",\"method\":\"eth_call\",\"params\":[{"
+            + "\"from\":\"0x0\", \"to\": \"0x0\"}, "
+            + "\"latest\","
+            + "{\""
+            + ADDRESS_HEX1
+            + "\":"
+            + "{"
+            + "\"balance\": \"0x01\","
+            + "\"nonce\": \""
+            + "0x9e"
+            + "\""
+            + "}}],\"id\":1}";
+
+    final JsonRpcRequestContext request = new JsonRpcRequestContext(readJsonAsJsonRpcRequest(json));
+    final AccountOverrideMap accountOverrideParam =
+        request.getRequiredParameter(2, AccountOverrideMap.class);
+
+    final AccountOverride accountOverride =
+        accountOverrideParam.get(Address.fromHexString(ADDRESS_HEX1));
+
+    assertThat(accountOverride.getBalance()).isEqualTo(Optional.of(Wei.of(1)));
+    assertThat(accountOverride.getNonce().get()).isEqualTo(158); // 0x9e
+    assertFalse(accountOverride.getStateDiff().isPresent());
+  }
+
+  @Test
   public void jsonWithStorageOverridesDeserializesCorrectly() throws Exception {
     final String json =
         "{\"jsonrpc\":\"2.0\",\"method\":\"eth_call\",\"params\":[{"
@@ -107,7 +135,7 @@ public class AccountOverrideParameterTest {
             + "\":"
             + "{"
             + "\"balance\": \"0x01\","
-            + "\"nonce\": 88,"
+            + "\"nonce\": \"0x9E\","
             + "\"stateDiff\": {"
             + "\""
             + STORAGE_KEY
@@ -124,7 +152,7 @@ public class AccountOverrideParameterTest {
 
     final AccountOverride accountOverride =
         accountOverrideParam.get(Address.fromHexString(ADDRESS_HEX1));
-    assertThat(accountOverride.getNonce()).isEqualTo(Optional.of(88L));
+    assertThat(accountOverride.getNonce().get()).isEqualTo(158);
 
     assertTrue(accountOverride.getStateDiff().isPresent());
     assertThat(accountOverride.getStateDiff().get().get(STORAGE_KEY)).isEqualTo(STORAGE_VALUE);
@@ -141,7 +169,7 @@ public class AccountOverrideParameterTest {
             + "\":"
             + "{"
             + "\"balance\": \"0x01\","
-            + "\"nonce\": 88,"
+            + "\"nonce\": \"0x9E\","
             + "\"stateDiff\": {"
             + "\""
             + STORAGE_KEY
@@ -154,7 +182,7 @@ public class AccountOverrideParameterTest {
             + "\":"
             + "{"
             + "\"balance\": \"0xFF\","
-            + "\"nonce\": 99,"
+            + "\"nonce\": \"0x9D\","
             + "\"stateDiff\": {"
             + "\""
             + STORAGE_KEY
@@ -171,14 +199,14 @@ public class AccountOverrideParameterTest {
 
     final AccountOverride accountOverride1 =
         accountOverrideParam.get(Address.fromHexString(ADDRESS_HEX1));
-    assertThat(accountOverride1.getNonce()).isEqualTo(Optional.of(88L));
+    assertThat(accountOverride1.getNonce().get()).isEqualTo(158);
     assertThat(accountOverride1.getBalance()).isEqualTo(Optional.of(Wei.fromHexString("0x01")));
     assertTrue(accountOverride1.getStateDiff().isPresent());
     assertThat(accountOverride1.getStateDiff().get().get(STORAGE_KEY)).isEqualTo(STORAGE_VALUE);
 
     final AccountOverride accountOverride2 =
         accountOverrideParam.get(Address.fromHexString(ADDRESS_HEX2));
-    assertThat(accountOverride2.getNonce()).isEqualTo(Optional.of(99L));
+    assertThat(accountOverride2.getNonce().get()).isEqualTo(157);
     assertThat(accountOverride2.getBalance()).isEqualTo(Optional.of(Wei.fromHexString("0xFF")));
     assertTrue(accountOverride2.getStateDiff().isPresent());
     assertThat(accountOverride2.getStateDiff().get().get(STORAGE_KEY)).isEqualTo(STORAGE_VALUE);

